lib/Sema/ConstraintSystem.cpp

```diff
@@ -1753,8 +1764,7 @@ static void validateInitializerRef(ConstraintSystem &cs, ConstructorDecl *init,
                               ->getInstanceType()
                               ->getWithoutParens();
       if (!cs.isTypeReference(baseExpr) && instanceType->isExistentialType()) {
-        (void)cs.recordFix(AllowInvalidInitRef::onProtocolMetatype(
-            cs, baseType, init, true, baseExpr->getSourceRange(), locator));
+        recordOnProtocolMetatypeFix(/*isStaticallyDerived*/ true);
```

Comment:
Sorry, not really what I meant, I was hopping that we could simplify the `if`s which leads to these fixes being produced...

Reply:
@xedin I have simplified the checks a more now, but I am not sure if it can be simplified any further.

Reply:
Ok, fair enough, that could be refactored later too! Let's get the rest of the regressed diagnostics under control (I think there is at least one which is related to super.init outside of initializer), and I think we should be ready to merge this PR.

Reply:
@xedin Okay! For that test, it doesn't get detected as `UR_InstanceMemberOnType` or `UR_TypeMemberOnInstance`. When I ran the tests, the unviable reasons were all different (most were type member on instance but others were label mismatch, etc).

Reply:
if the new logic doesn't get triggered why would that test case regress?

Reply:
@xedin Hmm, I think I am wrong. The fix does indeed get created, but the logic for emitting `could_not_use_value_member` is in `diagnoseUnviableLookupResults()` and not in the new logic 🤔 So it seems like the fix isn't getting diagnosed!

Reply:
Returning `SolutionKind::Error` in `simplifyMemberConstraint` as originally done if `!result.UnviableCandidates.empty()` works around the problem, but it's obviously not the correct thing to do as our fixes won't get created otherwise.

Reply:
I think it would make sense to either add new fix+diagnostic for this situation or incorporate the logic from CSDiag into `AllowTypeOrInstanceMemberFailure`.

Reply:
@xedin Hmm okay, I'll take a look tomorrow! Something's strange is going on as the fix is indeed being created, but not being diagnosed (i.e. nothing is calling `diagnoseAsError()` on it!). I am not sure how we're ending up in `diagnoseUnviableLookupResults()` if we've already recorded a fix.